architecture/2024-08-controller-component-threat-model.md

```diff
@@ -0,0 +1,186 @@
+3. **Controller Reconciliation**: When the action (create/update/delete) is validated by the Validating Webhook as discussed in the previous step, Kubernetes API reaches out to the Controller component and triggers the necessary action. Then, the controllers reconcile the resource's state to match the desired configuration.
+4. **Communication with UCPD**: Controller also calls UCPD in some cases when doing reconcilations. The communication with UCPD is one-way, meaning that UCPD doesn't make calls to the Controller component, just responds to its requests. Controller reaches out to UCPD to read a resource and/or write to a resource. In some cases, it can request secrets for a given resource.
+
+### Detailed Data Flows
+
+#### Kubernetes API Server - Validating Webhook for the Recipe Resource
+
+- **Assumptions**:
+  - TLS Certificates are defined on the startup of Controller component.
+- **Flow**:
+  1. Kubernetes API Server receives a request to create, update, or delete a Recipe or Radius Deployment resource.
+  1. Kubernetes API Server sends the request to the Validating Webhook for validation.
+  1. The webhook evaluates the request and responds with an approval or a rejection.
+  1. If approved, Kubernetes API Server proceeds with the action.
+- **Risks and Mitigations**:
+  - **Risk**: Could someone inject dangerous data into the resource object that would also get an approval from the webhook?
+    - **Mitigation**: Implement comprehensive validation logic in the webhook to check for malicious data patterns, enforce strict schema validation, and sanitize inputs.
+  - **Risk**: What if the webhook server is unavailable or slow to respond?
+    - **Mitigation**: Configure timeouts and retries for webhook calls, and implement fallback mechanisms to handle webhook failures gracefully.
+    - **References**:
+      - <https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#timeouts>
+      - <https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy>
+  - **Risk**: What if the TLS certificates used for securing communication are compromised?
+    - **Mitigation**: Implement strong certificate management practices, including regular rotation of certificates, using short-lived certificates, and monitoring for certificate anomalies.
+  - **Risk**: Could an insider with access to the webhook server modify its behavior to approve malicious requests?
+    - **Mitigation**: Implement strict access controls, audit logs, and monitoring for changes to the webhook server configuration and code.
+  - **Risk**: What if the webhook logic contains bugs that lead to incorrect validation decisions?
+    - **Mitigation**: Implement thorough testing, including unit tests, integration tests, and fuzz testing, to ensure the correctness of the webhook logic.
```

Comment:
What is the worst case scenario if this does happen?

Reply:
The webhook will be returning approvals instead of rejections and the reconciler will be trying to reconcile objects that shouldn't be reconciled. I think we had a similar issue a few months ago. Controller was getting in a failed state. I will share the issue if I can find it.